tests/plugins/test_settings.py

```diff
@@ -0,0 +1,135 @@
+
+#: Value for the string type parameter (used in test condarc below)
+STRING_PARAMETER_VALUE = "test_value"
+
+#: Value for the string type parameter (used in test condarc below)
+STRING_PARAMETER_ENV_VAR_VALUE = "env_var_value"
+
+#: condarc file with our test setting present
+CONDARC_TEST_ONE = f"""
+plugins:
+  {STRING_PARAMETER_NAME}: {STRING_PARAMETER_VALUE}
+"""
+
+string_parameter = PrimitiveParameter("", element_type=str)
+
+string_config_parameter = plugins.CondaSetting(
+    name=STRING_PARAMETER_NAME,
+    description="Test string type setting",
+    parameter=string_parameter,
+    aliases=(STRING_PARAMETER_ALIAS,),
+)
+
+
+class SettingPlugin:
+    @plugins.hookimpl
+    def conda_settings(self):
+        yield string_config_parameter
+
+
+@pytest.fixture()
+def setting_plugin_manager(
+    plugin_manager: CondaPluginManager,
+) -> CondaPluginManager:
+    """
+    Loads our ``SettingPlugin`` class using the ``plugin_manager`` fixture
+    """
+    plugin = SettingPlugin()
+    plugin_manager.register(plugin)
+
+    yield plugin_manager
+
+
+@pytest.fixture()
+def condarc_plugin_manager(setting_plugin_manager):
+    """
+    Update the context object to load our test condarc file containing a setting
+    defined by a plugin.
+    """
+    reset_context()
+    context._set_raw_data(
+        {
+            "testdata": YamlRawParameter.make_raw_parameters(
+                "testdata", yaml_round_trip_load(CONDARC_TEST_ONE)
+            )
+        }
+    )
+
+    try:
+        del context.plugins  # clear cached property
+    except AttributeError:
+        pass
+
+    return setting_plugin_manager
+
+
+def test_get_settings(setting_plugin_manager):
+    """
+    Ensure the settings method returns what we expect
+    """
+    config_params = setting_plugin_manager.get_settings()
+    assert len(config_params) == 1
+    assert config_params.get(STRING_PARAMETER_NAME) == (
+        string_parameter,
+        (STRING_PARAMETER_ALIAS,),
+    )
+
+
+def test_load_configuration_parameters(setting_plugin_manager):
+    """
+    Ensure that the setting is available via the context object
+    """
+    setting_plugin_manager.load_settings()
+    assert hasattr(context.plugins, STRING_PARAMETER_NAME)
+
+
+def test_load_plugin_settings_with_condarc(condarc_plugin_manager):
+    """
+    Ensure that when we define a custom plugin setting in a condarc
+    file that the value shows up on the context object.
+    """
+    assert getattr(context.plugins, STRING_PARAMETER_NAME) == STRING_PARAMETER_VALUE
+
+
+def test_load_plugin_config_with_env_var(
+    monkeypatch: MonkeyPatch, setting_plugin_manager
+):
+    """
+    Ensure that when an environment variable is set for a plugin setting
+    it is read correctly.
+    """
+    monkeypatch.setenv(
+        f"CONDA_PLUGINS_{STRING_PARAMETER_NAME.upper()}", STRING_PARAMETER_ENV_VAR_VALUE
+    )
+    reset_context()
+
+    try:
+        del context.plugins
+    except AttributeError:
```

Comment:
Actually `del context.plugins` will always raise an `AttributeError` and never do anything since `plugins` is a read-only `property`

Reply:
@schuylermartin45 @kenodegard,

"plugins" is a cached property and the according the Python documentation, deleting it is the way to clear the cache:

> The cached value can be cleared by deleting the attribute. This allows the cached_property method to run again.

(source: https://docs.python.org/3/library/functools.html#functools.cached_property)

That is wrapped in try/except block because sometimes this cache has not been set yet (i.e. the property still hasn't been accessed). Albeit, it looks a little ugly in the tests, but that was the best solution I could find to ensure the tests pass every time.

Reply:
From the same docs:

> The cached_property does not prevent a possible race condition in multi-threaded usage. The getter function could run more than once on the same instance, with the latest run setting the cached value. If the cached property is idempotent or otherwise not harmful to run more than once on an instance, this is fine. If synchronization is needed, implement the necessary locking inside the decorated getter function or around the cached property access.

So there's a race condition. Ideally our tests should be deterministic. I think we have a couple of options:
1) Make the cache thread safe with a synchronization mechanism
2) Use `@property` (I'm guessing that's a non-starter and there's a reason you're using the cached version)
3) Mock the cache. I'm only okay with this if we expect that the only way this property is being used in a multi-threaded way is during the unit testing. Otherwise the race condition will crop up under normal use.

Reply:
Similar to 3 I suppose you could also guarantee initialization order of the cache with a fixture.

Reply:
oh wow I did not know that `functools.cached_property` had implemented yet another way to clear the cache, I was expecting the same `cache_clear()` that `functools.lru_cache` has

I'd be in favor of moving the try-except into a fixture, I also think we should be clearing the `context.plugins` value for both setup and teardown

Reply:
@kenodegard, put this in a fixture with some more documentation.

@schuylermartin45, as we discussed, we are not worried at all about this running in a multi-threaded environment 😱. The `context` object is not thread safe!